lib/Frontend/CompilerInvocation.cpp

```diff
@@ -206,28 +199,27 @@ class ArgsToFrontendInputsConverter {
     }
     // Keep buffer alive because code passes around StringRefs.
     FilelistBuffer = std::move(*filelistBufferOrError);
+    bool hadDuplicates = false;
     for (auto file : llvm::make_range(llvm::line_iterator(*FilelistBuffer),
-                                      llvm::line_iterator())) {
-      addFile(file, Whence::SecondaryFromFileList);
-    }
-    return false;
+                                      llvm::line_iterator()))
+      hadDuplicates = addFile(file) || hadDuplicates;
+    return false; // Don't bail out for duplicates, too many tests depend on it.
   }
 
-  bool setPrimaryFiles() {
-    for (StringRef primaryFile : PrimaryFiles) {
-      const auto iterator = FileIndices.find(primaryFile);
-      // Catch "swiftc -frontend -c -filelist foo -primary-file
-      // some-file-not-in-foo".
-      if (iterator == FileIndices.end()) {
-        assert(FilelistPathArg != nullptr &&
-               "Missing primary with no filelist");
-        Diags.diagnose(SourceLoc(), diag::error_primary_file_not_found,
-                       primaryFile, filelistPath());
-        return true;
-      }
-      Inputs.addPrimaryInputFilename(iterator->second);
+  bool addFile(StringRef file) {
+    if (Files.count(file) == 0) {
+      Files.insert(file);
```

Comment:
Optimization nitpick: `insert` returns a boolean (as part of a pair) that will tell you if the entry was in the set already, which saves a hash lookup.

Reply:
Cool! Done.